### PR TITLE
Optionally allow redefinition of variable with different type

### PR DESCRIPTION
**Pull request by JukkaL**
If the flag `--allow-redefinition` is used, it's now possible to redefine a 
variable with a different type. For example, this code will be okay:

```
x = 0  # Define 'x' with type 'int'
print(x)
x = 'a'  # Define another 'x' with type 'str'
print(x)
```

The variable needs to be read before redefinition is possible. This is mainly 
to allow these idioms:

```
x: List[int]
x = []  # Not a redefinition

y = None  # type: List[int]  # Strict optional disabled
y = []  # Not a redefinition
```

It's also okay to use a type annotation in any of the assignments. This is
fine:

```
x: int = 0
print(x)
x: str = 'a'
print(x)
```

Redefinition is only allowed to happen in the same block and the nesting level
as the original definition. This lets us keep the implementation relatively simple,
and there will hopefully be less of a readability impact, since it's less likely that
users accidentally redefine a variable or use redefinition in confusing ways. We 
may want to lift the restriction in the future, however. Function arguments can
be redefined in the body but only in non-nested blocks.

The implementation does a simple static analysis to rename variables to make
them distinct when we detect redefinition. This happens before semantic 
analysis pass 1.

The new behavior is not fully backward compatible with the old behavior. For
example, here we lose type context in the second assignment:

```
x: List[int] = []
f(x)
x = []  # Need type annotation since this creates an independent variable
```

Internally we use a sequence of single quotes as a variable suffix
to represent renamed variants. The implementation provides an `unmangle()`
function to get the original name from a potentially mangled name.

It's still impossible to redefine final names, classes and certain other things.
The main use case is simple variables, either in single or multiple assignments.

I did some benchmarking (not with the latest version though) and the performance 
impact of the new pass was minor (under 2% IIRC).

Things to do in additional PRs:
* Add documentation.
* Make sure we unmangle names everywhere.
* Consider allowing redefinition to happen sometimes in another block.

### CODE REVIEW

File: test-data/unit/check-redefine.test [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+    y = '' # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+class C:
+    y = 0
+    y = '' # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+def g() -> None:
+    # _ is a special case
+    _ = 0
+    _ = ''
+    x, _ = 0, C()
+[builtins fixtures/tuple.pyi]
```
Comment (ilevkivskyi) [inline test-data/unit/check-redefine.test]: I would add test cases for
```python
e = 1
e
try:
    ...
except Exception as e:
    ...
```
and
```python
with open('file') as f:  # IO[str]
    ...
with open('file', 'b') as f:  # IO[bytes]
    ...
```
(maybe plus some variations).

File: mypy/renaming.py [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+
+    def visit_continue_stmt(self, stmt: ContinueStmt) -> None:
+        self.reject_redefinition_of_vars_in_loop()
+
+    def visit_try_stmt(self, stmt: TryStmt) -> None:
+        self.enter_with_or_try()
+        super().visit_try_stmt(stmt)
+        self.leave_with_or_try()
+
+    def visit_with_stmt(self, stmt: WithStmt) -> None:
```
[the remark this answers is not in the document]
Comment (JukkaL) [reply mypy/renaming.py]: Try/except has special handling in the type checker, so it's not needed (added comment). Fixed with statements.

File: mypy/renaming.py [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+            self.analyze_lvalue(lvalue.expr, is_nested=is_nested)
+
+    def visit_name_expr(self, expr: NameExpr) -> None:
+        self.handle_ref(expr)
+
+    # Helpers for renaming references
+
+    def handle_arg(self, name: str) -> None:
+        """Store function argument."""
+        if name not in self.refs[-1]:
```
[the remark this answers is not in the document]
Comment (JukkaL) [reply mypy/renaming.py]: Removed the check.

File: mypy/renaming.py [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+            names[-1].append(expr)
+        num_reads = self.num_reads[-1]
+        num_reads[name] = num_reads.get(name, 0) + 1
+
+    def flush_refs(self) -> None:
+        """Rename all references within the current scope.
+
+        This will be called at the end of a scope.
+        """
+        is_func = self.is_nested()
```
[the remark this answers is not in the document]
Comment (JukkaL) [reply mypy/renaming.py]: Fixed this.

File: test-data/unit/check-final.test [hunk 5]
```diff
@@ -336,14 +336,15 @@ class C:
...
 from typing import Final
 
 x: Final = 1
-x  # Dummy reference to allow renaming once implemented
+x
 x = 2  # E: Cannot assign to final name "x"
 def f() -> int:
     global x
-    x = 3  # E: Cannot assign to final name "x"
+    x = 3  # No error here is okay since we reported an error above
```
[the remark this answers is not in the document]
Comment (JukkaL) [reply test-data/unit/check-final.test]: Yes. Added a test case.

File: test-data/unit/check-redefine.test (hunk 1 above)
Comment (JukkaL) [reply test-data/unit/check-redefine.test]: Added test cases.

**Comment by JukkaL** [comment]
I think that I've now addressed all the feedback.

Agreed that this will need to be documented by the next release (or we should hide the option).

> Maybe when we detect UninhabitedType, instead of immediately giving an error we can try to re-infer type of r.h.s. using type of the "previous" x (with one prime less) as a context?

I thought about this but I couldn't come up with a simple implementation, so I decided to postpone this. Calculating the previous 'x' is more complicated, since the rules for renaming are different for globals and locals, for example. This is a potential future improvement. I'll create a follow-up issue about this.

**Comment by neighthan** [comment]
Is there a way to only allow redefinition if an explicit type is given? So that 

```python
x = 0
print(x)
x: str = 'a'
print(x)
```

is valid but

```python
x = 0
print(x)
x = 'a'
print(x)
```

is not? This would prevent any accidental redefinition and make it clear that the type is supposed to be changing (in the example above it's clear anyway, but if we had `x = some_func()` it might not be obvious anymore).

**Comment by oakkitten** [comment]
this doesn't seem to work for `def`:

```
from typing import Callable, cast

foo: str = "a"
foo: int = cast(int, foo)

bar: Callable[[], int] = lambda: 1
bar: int = cast(int, bar)

def baz() -> int:
    return 1
baz: int = cast(int, baz)
```
> 11: error: Name 'baz' already defined on line 9

any workarounds?

**Pull request merged**